binderhub/builder.py

```diff
@@ -425,33 +425,72 @@ async def get(self, provider_prefix, _unescaped_spec):
             ref_url=self.ref_url,
         )
 
-        self.build = build = BuildClass(
-            q=q,
-            # api object can be None if we are using FakeBuild
-            api=self.settings.get("kubernetes_client"),
-            name=build_name,
-            namespace=self.settings["build_namespace"],
-            repo_url=repo_url,
-            ref=ref,
-            image_name=image_name,
-            push_secret=push_secret,
-            build_image=self.settings["build_image"],
-            memory_limit=self.settings["build_memory_limit"],
-            memory_request=self.settings["build_memory_request"],
-            docker_host=self.settings["build_docker_host"],
-            node_selector=self.settings["build_node_selector"],
-            appendix=appendix,
-            log_tail_lines=self.settings["log_tail_lines"],
-            git_credentials=provider.git_credentials,
-            sticky_builds=self.settings["sticky_builds"],
-        )
+        if issubclass(BuildClass, Build):
+            # Deprecated
+            build = BuildClass(
+                q=q,
+                # api object can be None if we are using FakeBuild
+                api=self.settings.get("kubernetes_client"),
+                name=build_name,
+                namespace=self.settings["build_namespace"],
+                repo_url=repo_url,
+                ref=ref,
+                image_name=image_name,
+                push_secret=push_secret,
+                build_image=self.settings["build_image"],
+                memory_limit=self.settings["build_memory_limit"],
+                memory_request=self.settings["build_memory_request"],
+                docker_host=self.settings["build_docker_host"],
+                node_selector=self.settings["build_node_selector"],
+                appendix=appendix,
+                log_tail_lines=self.settings["log_tail_lines"],
+                git_credentials=provider.git_credentials,
+                sticky_builds=self.settings["sticky_builds"],
+            )
+        else:
+            build = BuildClass(
+                # Commented properties should be set in traitlets config
+                parent=self.settings["traitlets_parent"],
+                q=q,
+                name=build_name,
+                # namespace=self.settings["build_namespace"],
+                repo_url=repo_url,
+                ref=ref,
+                image_name=image_name,
+                # push_secret=push_secret,
+                # build_image=self.settings["build_image"],
+                # memory_limit=self.settings["build_memory_limit"],
+                # memory_request=self.settings["build_memory_request"],
+                # docker_host=self.settings["build_docker_host"],
+                # node_selector=self.settings["build_node_selector"],
+                appendix=appendix,
```

Comment:
Should appendix be passed in from `BinderHub.appendix` which therefore overrides `BuildExecutor.appendix`, or should it be set by the `BuildExecutor.appendix` Traitlet only? On one hand it's only applicable to repo2docker which points to it being configurable in `BuildExecutor`, on the other hand changes to appendix can result in major changes to the built images which implies it's a higher-level BinderHub feature.

Reply:
In general, I think it's best for configurables to only show up one place, and close to the place they are used - there are likely a few things on BinderHub that should be on BuildExecutor now that it's configurable. We probably shouldn't have things that are set on BinderHub, but only to be used on BuildExecutor. That organized config a bit better, I think, and eliminates the App->settings->Build pipeline. But it's a backward-compatibility issue to move config.

Reply:
I've changed it so `BinderHub.appendix` is ignored by `BuildExecutor`.

Note `BinderHub.appendix` isn't used directly, it's expanded as a Jinja2 template:
https://github.com/jupyterhub/binderhub/blob/b78cac34ee6a2324822d9596137462b022d7dec9/binderhub/builder.py#L420-L426

The appendix is empty by default so in the default case this has no effect. In the past `appendix` was set in mybinder.org-deploy, but that was subsequently removed in https://github.com/jupyterhub/mybinder.org-deploy/pull/1284/ , the relevant information is now passed at runtime, so this should be OK to remove in a later breaking PR.